### ISSUE TITLE
TypeError in plan monitor when using constant objects in preconditions or effects

### ISSUE THREAD

**Issue by marcvinci** [aiplan4eu/embedded-systems-bridge#39]
## Error:
When using a constant object in a precondition or an effect of an UP action, the monitoring raises the following TypeError:
```
Traceback (most recent call last):
  File "robot_example_constant.py", line 92, in <module>
    main()
  File "robot_example_constant.py", line 83, in main
    dispatcher.execute_plan(plan, graph_executor)
  File "/home/mvinci/.local/lib/python3.8/site-packages/up_esb/plexmo/dispatcher.py", line 123, in execute_plan
    result = self._executor.execute_action(node_id)
  File "/home/mvinci/.local/lib/python3.8/site-packages/up_esb/execution/executor.py", line 99, in execute_action
    result = self._check_preconditions(task_id)
  File "/home/mvinci/.local/lib/python3.8/site-packages/up_esb/execution/executor.py", line 168, in _check_preconditions
    compile(condition, filename="<ast>", mode="eval"), self._context
TypeError: got an invalid type in Constant: Object
```
## Code Snippet:
An example code snippet, which produces the error, can be seen here:
```
import unified_planning as up
from unified_planning.shortcuts import Not

from up_esb.bridge import Bridge
from up_esb.plexmo import PlanDispatcher

class Location:
    def __init__(self, name):
        self.name = name

    def __eq__(self, __value: object) -> bool:
        if isinstance(__value, Location):
            return self.name == __value.name
        return False

class Robot:
    location = Location("base_station")

    @classmethod
    def move_to_l1(cls, l_from: Location):
        Robot.location = Location("l1")
        return True

def robot_at_fun(l: Location):
    return Robot.location == l

def define_problem():
    bridge = Bridge()

    bridge.create_types([Location, Robot])

    robot_at = bridge.create_fluent_from_function(robot_at_fun)

    robot = bridge.create_object("r1", Robot())
    l1 = bridge.create_object("l1", Location("l1")) #<------- the constant object to be used in action move_to_l1
    base_station = bridge.create_object("base_station", Robot.location)

    move_to_l1, [l_from] = bridge.create_action(
        "move_to_l1", _callable=Robot.move_to_l1, l_from=Location
    )
    move_to_l1.add_precondition(robot_at(l_from))
    move_to_l1.add_precondition(Not(robot_at(l1))) #<----------------------- Using constant object l1 here
    move_to_l1.add_effect(robot_at(l_from), False)
    move_to_l1.add_effect(robot_at(l1), True) #<---------------------------- and here

    problem = bridge.define_problem()
    bridge.set_initial_values(problem)

    problem.set_initial_value(robot_at(base_station), True)

    problem.add_goal(robot_at(l1))

    return problem, bridge

def main():
    up.shortcuts.get_environment().credits_stream = None
    problem, bridge = define_problem()
    dispatcher = PlanDispatcher()

    plan = bridge.solve(problem, optimize_with_default_metric=False)
    graph_executor = bridge.get_executable_graph(plan)
    dispatcher.execute_plan(plan, graph_executor)

if __name__ == "__main__":
    main()
```

This seems to be an error from AST or how the AST expressions are built during the creation of the dependency graph.

**Comment by franklinselva** [comment]
@marcvinci Thanks for reporting the issue. I will take a look and come back to you.

**Comment by franklinselva** [comment]
Hi @marcvinci 

From your action definition, 

``` py
    move_to_l1, [l_from] = bridge.create_action(
        "move_to_l1", _callable=Robot.move_to_l1, l_from=Location
    )
    move_to_l1.add_precondition(robot_at(l_from))
    move_to_l1.add_precondition(Not(robot_at(l1))) #<----------------------- Using constant object l1 here
    move_to_l1.add_effect(robot_at(l_from), False)
    move_to_l1.add_effect(robot_at(l1), True) #<---------------------------- and here
```

The precondition `Not(robot_at(l1)` cannot be deduced to the expression that needs to be evaluated since it already returns a boolean value when the action is defined. You may need to find a different way to represent the condition.

But to evaluate if the effects are valid and succeeded, for your example, we need a lookup table to store the current state of the system which unfortunately is not available in the bridge at the moment. 

I will look into the possibilities if it can be added soon but I cannot promise that yet.

**Comment by marcvinci** [comment]
Hi @franklinselva,
I tried to fix the issue with this commit a855ad0. Would be nice if you can take a look and see if that works as a solution.

**Comment by franklinselva** [comment]
@marcvinci Thanks for the changes. It looks good to me. I added a few tests covering the changes you made.

I also stumbled upon a bug related to fluent type when looking into your changes. But it looks like it's out of scope for this issue.

You can merge the changes.
